packages/statemanager/src/flat/snapshot.ts

```diff
@@ -0,0 +1,277 @@
+export const CODE_PREFIX: Uint8Array = hexToBytes('0x' + '22')
+
+const { debug: createDebugLogger } = debugDefault
+
+type SnapshotElement = {
+  data: Uint8Array | undefined
+}
+
+function concatenateUint8Arrays(arrays: Uint8Array[]) {
+  const l = arrays.reduce((prev, curr) => prev + curr.length, 0)
+  const concatenatedArray = new Uint8Array(l)
+
+  let offset = 0
+  for (const array of arrays) {
+    concatenatedArray.set(array, offset)
+    offset += array.length
+  }
+
+  return concatenatedArray
+}
+
+export class Snapshot {
+  _db: LevelDB<Uint8Array, Uint8Array>
+  _debug: Debugger
+  _checkpoints = 0
+
+  /**
+   * Diff cache collecting the state of the cache
+   * at the beginning of checkpoint height
+   * (respectively: before a first modification)
+   *
+   * If the whole cache element is undefined (in contrast
+   * to the account), the element didn't exist in the cache
+   * before.
+   */
+  _diffCache: Map<string, SnapshotElement | undefined>[] = []
+  constructor(db?: LevelDB<Uint8Array, Uint8Array>) {
+    this._db = db ?? new LevelDB<Uint8Array, Uint8Array>()
+    this._diffCache.push(new Map<string, SnapshotElement | undefined>())
+    this._debug = createDebugLogger('client:snapshot')
+  }
+
+  async _saveCachePreState(key: Uint8Array) {
+    const keyHex = bytesToHex(key)
+    const diffMap = this._diffCache[this._checkpoints]
+    if (!diffMap.has(keyHex)) {
+      const oldElem: SnapshotElement | undefined = {
+        data: (await this._db.get(key)) as Uint8Array | undefined,
+      }
+      diffMap.set(keyHex, oldElem)
+    }
+  }
+
+  async putAccount(address: Address, account: Account): Promise<void> {
+    const key = concatenateUint8Arrays([ACCOUNT_PREFIX, keccak256(address.bytes)])
+    await this._saveCachePreState(key)
+
+    const value = account.serialize()
+    await this._db.put(key, value)
+  }
+
+  async getAccount(address: Address): Promise<Uint8Array | undefined> {
+    const key = concatenateUint8Arrays([ACCOUNT_PREFIX, keccak256(address.bytes)])
+
+    return this._db.get(key)
+  }
+
+  _getAccounts(): Promise<[Uint8Array, Uint8Array | undefined][]> {
+    const prefix = ACCOUNT_PREFIX
+    return this._db.byPrefix(prefix, { keyEncoding: KeyEncoding.Bytes })
+  }
+
+  async delAccount(address: Address): Promise<void> {
+    const key = concatenateUint8Arrays([ACCOUNT_PREFIX, keccak256(address.bytes)])
+    await this._saveCachePreState(key)
+
+    return this._db.del(key)
+  }
+
+  /**
+   * Deletes not only the account itself, but also the code
+   * and any storage items if available.
+   */
+  async clearAccount(address: Address): Promise<void> {
+    const rawAccount = await this.getAccount(address)
+    if (rawAccount === undefined) return
+
+    await this.delAccount(address)
+
+    const account = Account.fromRlpSerializedAccount(rawAccount)
+    if (!equalsBytes(account.codeHash, KECCAK256_NULL)) {
+      // await this.delCode(account.codeHash)
```

Comment:
Why is this commented out?  Are we optimistically thinking other accounts might have the same code deposited?

Reply:
Yes, that was my main concern, so I have it commented out until I can uncomment or remove it with confidence. I'll add a TODO comment above it as well, for clarity.